File: packages/dart/lib/src/kenya_locations.dart
import 'dart:math' as math;

import 'package:kenya_locations/src/src.dart';

/// Provides access to Kenya's administrative and geographical location data.
///
/// Includes methods for retrieving counties, sub-counties, constituencies,
/// wards, localities, and areas, as well as searching across supported
/// location types.
///
/// Search supports case-insensitive substring matching and typo-tolerant
/// matching using Levenshtein distance.
class KenyaLocations {
  static final List<County> _counties = rawCounties
      .map(County.fromMap)
      .toList();

  static final List<SubCounty> _subCounties = rawSubCounties
      .map(SubCounty.fromMap)
      .toList();

  static final List<Constituency> _constituencies = rawConstituencies
      .map(Constituency.fromMap)
      .toList();

  static final List<Ward> _wards = rawWards.map(Ward.fromMap).toList();

  static final List<Locality> _localities = rawLocalities
      .map(Locality.fromMap)
      .toList();

  static final List<Area> _areas = rawAreas.map(Area.fromMap).toList();

  /// Returns all counties in Kenya.
  static List<County> getCounties() => _counties;

  /// Returns the county with the given administrative code.
  ///
  /// County codes are represented as three-digit strings, for example `047`
  /// for Nairobi County.
  ///
  /// Returns `null` if no county matches the [code].
  static County? getCountyByCode(String code) =>
      _counties.where((c) => c.code == code).firstOrNull;

  /// Returns the county with the given name.
  ///
  /// The comparison is case-insensitive. Returns `null` if no county matches
  /// the [name].
  static County? getCountyByName(String name) => _counties
      .where((c) => c.name.toLowerCase() == name.toLowerCase())
      .firstOrNull;

  /// Returns all sub-counties in Kenya.
  static List<SubCounty> getSubCounties() => _subCounties;

  /// Returns all sub-counties belonging to the given county.
  ///
  /// The comparison is case-insensitive.
  static List<SubCounty> getSubCountiesInCounty(String countyName) =>
      _subCounties
          .where((s) => s.county.toLowerCase() == countyName.toLowerCase())
          .toList();

  /// Returns all constituencies in Kenya.
  static List<Constituency> getConstituencies() => _constituencies;

  /// Returns all constituencies belonging to the given county.
  ///
  /// The comparison is case-insensitive.
  static List<Constituency> getConstituenciesInCounty(String countyName) =>
      _constituencies
          .where((c) => c.county.toLowerCase() == countyName.toLowerCase())
          .toList();

  /// Returns all wards in Kenya.
  static List<Ward> getWards() => _wards;

  /// Returns all wards belonging to the given constituency.
  ///
  /// The comparison is case-insensitive.
  static List<Ward> getWardsInConstituency(String constituencyName) => _wards
      .where(
        (w) => w.constituency.toLowerCase() == constituencyName.toLowerCase(),
      )
      .toList();

  /// Returns the constituency containing the given ward.
  ///
  /// The comparison of the ward name is case-insensitive. Returns `null` if
  /// the ward or its parent constituency cannot be found.
  static Constituency? getConstituencyOfWard(String wardName) {
    final ward = _wards
        .where((w) => w.name.toLowerCase() == wardName.toLowerCase())
        .firstOrNull;
    if (ward == null) return null;

    return _constituencies
        .where((c) => c.name == ward.constituency)
        .firstOrNull;
  }

  /// Returns all localities in Kenya.
  static List<Locality> getLocalities() => _localities;

  /// Returns all localities belonging to the given county.
  ///
  /// The comparison is case-insensitive.
  static List<Locality> getLocalitiesInCounty(String countyName) => _localities
      .where((l) => l.county.toLowerCase() == countyName.toLowerCase())
      .toList();

  /// Returns all areas in Kenya.
  static List<Area> getAreas() => _areas;

  /// Returns all areas belonging to the given locality.
  ///
  /// The comparison is case-insensitive.
  static List<Area> getAreasInLocality(String localityName) => _areas
      .where((a) => a.locality.toLowerCase() == localityName.toLowerCase())
      .toList();

  /// Searches across counties, wards, localities, and areas.
  ///
  /// The search is case-insensitive and supports typo-tolerant matching.
  /// Exact substring matches are ranked first, followed by fuzzy matches
  /// based on Levenshtein distance.
  ///
  /// A fuzzy match is accepted when its edit distance is within 40% of the
  /// query length. This allows common spelling mistakes, character
  /// substitutions, insertions, and deletions.
  ///
  /// Queries shorter than two characters return no results.
  ///
  /// Results are ordered from the closest match to the least similar match
  /// and limited to [limit] results.
  static List<SearchResult<dynamic>> search(String query, {int limit = 10}) {
    final q = query.trim();

    if (q.length < 2 || limit <= 0) return [];

    final scored = <({double score, SearchResult<dynamic> result})>[];

    void collect<T>(
      List<T> items,
      SearchType type,
      String Function(T item) nameOf,
    ) {
      for (final item in items) {
        final score = _fuzzyScore(q, nameOf(item));

        if (score != null) {
          scored.add((
            score: score,
            result: SearchResult(type: type, item: item),
          ));
        }
      }
    }

    collect(_counties, SearchType.county, (item) => item.name);
    collect(_wards, SearchType.ward, (item) => item.name);
    collect(_localities, SearchType.locality, (item) => item.name);
    collect(_areas, SearchType.area, (item) => item.name);

    scored.sort((a, b) => a.score.compareTo(b.score));

    return scored.take(limit).map((entry) => entry.result).toList();
  }
}

/// Calculates a fuzzy match score between [pattern] and [text].
///
/// Returns `0` for an exact substring match. Otherwise, compares the pattern
/// against windows of [text] using Levenshtein distance.
///
/// A lower score indicates a better match. Returns `null` when the text does
/// not fall within the allowed typo tolerance.
double? _fuzzyScore(String pattern, String text) {
  final p = pattern.toLowerCase();
  final t = text.toLowerCase();

  if (t.contains(p)) return 0;
  if (p.length < 2 || t.isEmpty) return null;

  final maxErrors = math.max(1, (p.length * 0.4).floor());
  final windowSize = p.length + maxErrors;
  var best = double.infinity;

  for (var start = 0; start < t.length; start++) {
    final end = math.min(start + windowSize, t.length);
    final window = t.substring(start, end);
    final distance = _levenshtein(p, window);

    if (distance <= maxErrors) {
      final score = distance / p.length;

      if (score < best) {
        best = score;
      }
    }

    if (best == 0) break;
  }

  return best <= 0.4 ? best : null;
}

/// Calculates the Levenshtein distance between two strings.
///
/// The distance represents the minimum number of insertions, deletions, and
/// substitutions required to transform [s] into [t].
int _levenshtein(String s, String t) {
  if (s.isEmpty) return t.length;
  if (t.isEmpty) return s.length;

  final dp = List<int>.generate(t.length + 1, (i) => i);

  for (var i = 1; i <= s.length; i++) {
    var previous = dp[0];
    dp[0] = i;

    for (var j = 1; j <= t.length; j++) {
      final current = dp[j];

      dp[j] = s[i - 1] == t[j - 1]
          ? previous
          : 1 + math.min(previous, math.min(dp[j], dp[j - 1]));

      previous = current;
    }
  }

  return dp[t.length];
}

File: packages/php/src/KenyaLocations.php
<?php

declare(strict_types=1);

namespace KenyaLocations;

/**
 * Kenyan administrative divisions loaded from the shared JSON dataset.
 *
 * No initialisation is required — call the static methods directly.
 */
final class KenyaLocations
{
    private function __construct()
    {
    }

    /** @return list<County> */
    public static function getCounties(): array
    {
        return Dataset::counties();
    }

    public static function getCountyByCode(string $code): ?County
    {
        foreach (Dataset::counties() as $county) {
            if ($county->code === $code) {
                return $county;
            }
        }

        return null;
    }

    public static function getCountyByName(string $name): ?County
    {
        $needle = self::normalise($name);

        foreach (Dataset::counties() as $county) {
            if (self::normalise($county->name) === $needle) {
                return $county;
            }
        }

        return null;
    }

    /** @return list<SubCounty> */
    public static function getSubCounties(): array
    {
        return Dataset::subCounties();
    }

    public static function getSubCountyByCode(string $code): ?SubCounty
    {
        foreach (Dataset::subCounties() as $subCounty) {
            if ($subCounty->code === $code) {
                return $subCounty;
            }
        }

        return null;
    }

    public static function getSubCountyByName(string $name): ?SubCounty
    {
        $needle = self::normalise($name);

        foreach (Dataset::subCounties() as $subCounty) {
            if (self::normalise($subCounty->name) === $needle) {
                return $subCounty;
            }
        }

        return null;
    }

    /** @return list<SubCounty> */
    public static function getSubCountiesInCounty(string $nameOrCode): array
    {
        $county = self::resolveCounty($nameOrCode);
        if ($county === null) {
            return [];
        }

        $needle = self::normalise($county->name);

        return array_values(array_filter(
            Dataset::subCounties(),
            static fn (SubCounty $item): bool => self::normalise($item->county) === $needle,
        ));
    }

    public static function getCountyOfSubCounty(string $nameOrCode): ?County
    {
        $subCounty = self::resolveSubCounty($nameOrCode);

        return $subCounty === null ? null : self::resolveCounty($subCounty->county);
    }

    /** @return list<Constituency> */
    public static function getConstituencies(): array
    {
        return Dataset::constituencies();
    }

    public static function getConstituencyByCode(string $code): ?Constituency
    {
        foreach (Dataset::constituencies() as $constituency) {
            if ($constituency->code === $code) {
                return $constituency;
            }
        }

        return null;
    }

    public static function getConstituencyByName(string $name): ?Constituency
    {
        $needle = self::normalise($name);

        foreach (Dataset::constituencies() as $constituency) {
            if (self::normalise($constituency->name) === $needle) {
                return $constituency;
            }
        }

        return null;
    }

    /** @return list<Constituency> */
    public static function getConstituenciesInCounty(string $nameOrCode): array
    {
        $county = self::resolveCounty($nameOrCode);
        if ($county === null) {
            return [];
        }

        $needle = self::normalise($county->name);

        return array_values(array_filter(
            Dataset::constituencies(),
            static fn (Constituency $item): bool => self::normalise($item->county) === $needle,
        ));
    }

    public static function getCountyOfConstituency(string $nameOrCode): ?County
    {
        $constituency = self::resolveConstituency($nameOrCode);

        return $constituency === null ? null : self::resolveCounty($constituency->county);
    }

    /** @return list<Ward> */
    public static function getWards(): array
    {
        return Dataset::wards();
    }

    public static function getWardByCode(string $code): ?Ward
    {
        foreach (Dataset::wards() as $ward) {
            if ($ward->code === $code) {
                return $ward;
            }
        }

        return null;
    }

    public static function getWardByName(string $name): ?Ward
    {
        $needle = self::normalise($name);

        foreach (Dataset::wards() as $ward) {
            if (self::normalise($ward->name) === $needle) {
                return $ward;
            }
        }

        return null;
    }

    /** @return list<Ward> */
    public static function getWardsInConstituency(string $nameOrCode): array
    {
        $constituency = self::resolveConstituency($nameOrCode);
        if ($constituency === null) {
            return [];
        }

        $needle = self::normalise($constituency->name);

        return array_values(array_filter(
            Dataset::wards(),
            static fn (Ward $item): bool => self::normalise($item->constituency) === $needle,
        ));
    }

    /** @return list<Ward> */
    public static function getWardsInCounty(string $nameOrCode): array
    {
        $names = [];
        foreach (self::getConstituenciesInCounty($nameOrCode) as $constituency) {
            $names[self::normalise($constituency->name)] = true;
        }

        return array_values(array_filter(
            Dataset::wards(),
            static fn (Ward $item): bool => isset($names[self::normalise($item->constituency)]),
        ));
    }

    /**
     * Wards whose constituency name matches the sub-county.
     *
     * `$nameOrCode` may be a sub-county administrative code or name.
     *
     * @return list<Ward>
     */
    public static function getWardsInSubCounty(string $nameOrCode): array
    {
        $subCounty = self::resolveSubCounty($nameOrCode);
        if ($subCounty === null) {
            return [];
        }

        $needle = self::normalise($subCounty->name);

        return array_values(array_filter(
            Dataset::wards(),
            static fn (Ward $item): bool => self::normalise($item->constituency) === $needle,
        ));
    }

    public static function getCountyOfWard(string $wardNameOrCode): ?County
    {
        $ward = self::findUniqueWard($wardNameOrCode);
        if ($ward === null) {
            return null;
        }

        $constituency = self::constituencyOf($ward);

        return $constituency === null ? null : self::resolveCounty($constituency->county);
    }

    public static function getSubCountyOfWard(string $wardNameOrCode): ?SubCounty
    {
        $ward = self::findUniqueWard($wardNameOrCode);

        return $ward === null ? null : self::resolveSubCounty($ward->constituency);
    }

    /**
     * Parent constituency of a ward.
     *
     * `$wardNameOrCode` may be an administrative code (exact) or a ward name
     * (case-insensitive). Codes win when both could match. Returns null when
     * the ward is missing, the name is reused (for example Township), or the
     * parent constituency cannot be found.
     */
    public static function getConstituencyOfWard(string $wardNameOrCode): ?Constituency
    {
        $ward = self::findUniqueWard($wardNameOrCode);

        return $ward === null ? null : self::constituencyOf($ward);
    }

    /** @return list<Locality> */
    public static function getLocalities(): array
    {
        return Dataset::localities();
    }

    public static function getLocalityByName(string $name): ?Locality
    {
        return self::getLocality($name);
    }

    /**
     * Locality by name, optionally scoped to a county when the name is reused.
     */
    public static function getLocality(string $name, ?string $countyName = null): ?Locality
    {
        $needle = self::normalise($name);

        if ($countyName !== null) {
            foreach (self::getLocalitiesInCounty($countyName) as $locality) {
                if (self::normalise($locality->name) === $needle) {
                    return $locality;
                }
            }

            return null;
        }

        foreach (Dataset::localities() as $locality) {
            if (self::normalise($locality->name) === $needle) {
                return $locality;
            }
        }

        return null;
    }

    /**
     * All localities matching a name across counties.
     *
     * @return list<Locality>
     */
    public static function getLocalitiesByName(string $name): array
    {
        $needle = self::normalise($name);

        return array_values(array_filter(
            Dataset::localities(),
            static fn (Locality $item): bool => self::normalise($item->name) === $needle,
        ));
    }

    /** @return list<Locality> */
    public static function getLocalitiesInCounty(string $nameOrCode): array
    {
        $county = self::resolveCounty($nameOrCode);
        if ($county === null) {
            return [];
        }

        $needle = self::normalise($county->name);

        return array_values(array_filter(
            Dataset::localities(),
            static fn (Locality $item): bool => self::normalise($item->county) === $needle,
        ));
    }

    public static function getCountyOfLocality(string $localityName): ?County
    {
        $locality = self::getLocalityByName($localityName);

        return $locality === null ? null : self::resolveCounty($locality->county);
    }

    /** @return list<Area> */
    public static function getAreas(): array
    {
        return Dataset::areas();
    }

    public static function getAreaByName(string $name): ?Area
    {
        $matches = self::getAreasByName($name);

        return $matches[0] ?? null;
    }

    /**
     * All areas matching a name across localities.
     *
     * @return list<Area>
     */
    public static function getAreasByName(string $name): array
    {
        $needle = self::normalise($name);

        return array_values(array_filter(
            Dataset::areas(),
            static fn (Area $item): bool => self::normalise($item->name) === $needle,
        ));
    }

    /** @return list<Area> */
    public static function getAreasInLocality(string $localityName): array
    {
        $needle = self::normalise($localityName);

        return array_values(array_filter(
            Dataset::areas(),
            static fn (Area $item): bool => self::normalise($item->locality) === $needle,
        ));
    }

    /** @return list<Area> */
    public static function getAreasInCounty(string $nameOrCode): array
    {
        $county = self::resolveCounty($nameOrCode);
        if ($county === null) {
            return [];
        }

        $needle = self::normalise($county->name);

        return array_values(array_filter(
            Dataset::areas(),
            static fn (Area $item): bool => self::normalise($item->county) === $needle,
        ));
    }

    public static function getCountyOfArea(string $areaName): ?County
    {
        $area = self::getAreaByName($areaName);

        return $area === null ? null : self::resolveCounty($area->county);
    }

    public static function getLocalityOfArea(string $areaName): ?Locality
    {
        $area = self::getAreaByName($areaName);
        if ($area === null) {
            return null;
        }

        return self::getLocality($area->locality, $area->county);
    }

    /**
     * Fuzzy search across counties, sub-counties, constituencies, wards,
     * localities, and areas.
     *
     * Exact substring matches rank first, then typo-tolerant Levenshtein
     * matches (edit distance within 40% of the query length). Queries shorter
     * than two characters return no results.
     *
     * @param list<SearchType>|null $types
     * @return list<SearchResult>
     */
    public static function search(string $query, int $limit = 20, ?array $types = null): array
    {
        $q = trim($query);
        if (mb_strlen($q) < 2 || $limit <= 0) {
            return [];
        }

        $allowed = $types === null ? null : array_fill_keys(
            array_map(static fn (SearchType $type): string => $type->value, $types),
            true,
        );

        /** @var list<array{score: float, result: SearchResult}> $scored */
        $scored = [];

        $collect = static function (array $items, SearchType $type) use ($q, $allowed, &$scored): void {
            if ($allowed !== null && !isset($allowed[$type->value])) {
                return;
            }

            foreach ($items as $item) {
                $score = self::fuzzyScore($q, $item->name);
                if ($score !== null) {
                    $scored[] = [
                        'score' => $score,
                        'result' => new SearchResult($type, $item),
                    ];
                }
            }
        };

        $collect(Dataset::counties(), SearchType::County);
        $collect(Dataset::subCounties(), SearchType::SubCounty);
        $collect(Dataset::constituencies(), SearchType::Constituency);
        $collect(Dataset::wards(), SearchType::Ward);
        $collect(Dataset::localities(), SearchType::Locality);
        $collect(Dataset::areas(), SearchType::Area);

        usort($scored, static fn (array $a, array $b): int => $a['score'] <=> $b['score']);

        return array_map(
            static fn (array $entry): SearchResult => $entry['result'],
            array_slice($scored, 0, $limit),
        );
    }

    /**
     * @return list<SearchResult>
     */
    public static function searchByType(string $query, SearchType $type, int $limit = 20): array
    {
        return self::search($query, $limit, [$type]);
    }

    private static function constituencyOf(Ward $ward): ?Constituency
    {
        return self::resolveConstituency($ward->constituency);
    }

    private static function resolveCounty(string $nameOrCode): ?County
    {
        $key = trim($nameOrCode);
        if ($key === '') {
            return null;
        }

        foreach (Dataset::counties() as $county) {
            if ($county->code === $key) {
                return $county;
            }
        }

        $needle = self::normalise($key);
        foreach (Dataset::counties() as $county) {
            if (self::normalise($county->name) === $needle) {
                return $county;
            }
        }

        return null;
    }

    private static function resolveConstituency(string $nameOrCode): ?Constituency
    {
        $key = trim($nameOrCode);
        if ($key === '') {
            return null;
        }

        foreach (Dataset::constituencies() as $constituency) {
            if ($constituency->code === $key) {
                return $constituency;
            }
        }

        $needle = self::normalise($key);
        foreach (Dataset::constituencies() as $constituency) {
            if (self::normalise($constituency->name) === $needle) {
                return $constituency;
            }
        }

        return null;
    }

    private static function resolveSubCounty(string $nameOrCode): ?SubCounty
    {
        $key = trim($nameOrCode);
        if ($key === '') {
            return null;
        }

        foreach (Dataset::subCounties() as $subCounty) {
            if ($subCounty->code === $key) {
                return $subCounty;
            }
        }

        $needle = self::normalise($key);
        foreach (Dataset::subCounties() as $subCounty) {
            if (self::normalise($subCounty->name) === $needle) {
                return $subCounty;
            }
        }

        return null;
    }

    /**
     * Unique ward by administrative code (exact) or name (case-insensitive).
     * Codes win. Returns null when the name is reused or nothing matches.
     */
    private static function findUniqueWard(string $wardNameOrCode): ?Ward
    {
        $key = trim($wardNameOrCode);
        if ($key === '') {
            return null;
        }

        $byCode = array_values(array_filter(
            Dataset::wards(),
            static fn (Ward $ward): bool => $ward->code === $key,
        ));
        if (count($byCode) === 1) {
            return $byCode[0];
        }

        $needle = self::normalise($key);
        $byName = array_values(array_filter(
            Dataset::wards(),
            static fn (Ward $ward): bool => self::normalise($ward->name) === $needle,
        ));

        return count($byName) === 1 ? $byName[0] : null;
    }

    /**
     * Relevance score in [0.0, 1.0). 0.0 is an exact substring match.
     */
    private static function fuzzyScore(string $pattern, string $text): ?float
    {
        $p = mb_strtolower($pattern);
        $t = mb_strtolower($text);

        if (str_contains($t, $p)) {
            return 0.0;
        }

        $pLen = mb_strlen($p);
        $tLen = mb_strlen($t);
        if ($pLen < 2 || $tLen === 0) {
            return null;
        }

        $maxErrors = max(1, (int) floor($pLen * 0.4));
        $windowSize = $pLen + $maxErrors;
        $best = INF;

        for ($start = 0; $start < $tLen; $start++) {
            $end = min($start + $windowSize, $tLen);
            $window = mb_substr($t, $start, $end - $start);
            $distance = self::levenshtein($p, $window);

            if ($distance <= $maxErrors) {
                $score = $distance / $pLen;
                if ($score < $best) {
                    $best = $score;
                }
            }

            if ($best === 0.0) {
                break;
            }
        }

        return $best <= 0.4 ? $best : null;
    }

    private static function levenshtein(string $s, string $t): int
    {
        $sChars = self::characters($s);
        $tChars = self::characters($t);
        $m = count($sChars);
        $n = count($tChars);

        if ($m === 0) {
            return $n;
        }
        if ($n === 0) {
            return $m;
        }

        $dp = range(0, $n);

        for ($i = 1; $i <= $m; $i++) {
            $previous = $dp[0];
            $dp[0] = $i;

            for ($j = 1; $j <= $n; $j++) {
                $current = $dp[$j];
                $dp[$j] = $sChars[$i - 1] === $tChars[$j - 1]
                    ? $previous
                    : 1 + min($previous, $dp[$j], $dp[$j - 1]);
                $previous = $current;
            }
        }

        return $dp[$n];
    }

    /**
     * @return list<string>
     */
    private static function characters(string $value): array
    {
        return preg_split('//u', $value, -1, PREG_SPLIT_NO_EMPTY) ?: [];
    }

    private static function normalise(string $value): string
    {
        return mb_strtolower(trim($value));
    }
}
